pkg/docker/config/config.go

```diff
@@ -0,0 +1,271 @@
+		return "", errors.Wrapf(err, "error opening file %q", path)
+	}
+	defer file.Close()
+
+	newData, err := json.MarshalIndent(auths, "", "\t")
+	if err != nil {
+		return "", errors.Wrapf(err, "error marshaling JSON %q", path)
+	}
+
+	if err = ioutil.WriteFile(path, newData, 0755); err != nil {
+		return "", errors.Wrapf(err, "error writing to file %q", path)
+	}
+	return registry, nil
+}
+
+// getPath gets the path of the auth.json file since the user can set their own path
+// by setting CRIO_AUTH_PATH
+// If CRIO_AUTH_PATH is not set, the default path is ~/.config/containers/auth.json
+func getPathToAuth() (string, error) {
+	if os.Getenv("CRIO_AUTH_PATH") != "" {
+		return filepath.Join(os.Getenv("CRIO_AUTH_PATH"), authCfgFileName), nil
+	}
+	dir := filepath.Join(homedir.Get(), authCfg)
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		if err = os.Mkdir(dir, 0600); err != nil {
+			return "", errors.Wrapf(err, "error creating directory %q", dir)
+		}
+	}
+	return filepath.Join(dir, authCfgFileName), nil
+}
+
+// readJSONFile unmarshals the authentications stored in the auth.json file
+func readJSONFile(path string) (dockerConfigFile, error) {
+	file, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0755)
+	if err != nil {
+		return dockerConfigFile{}, errors.Wrapf(err, "error creating/opening file %q", path)
+	}
+	defer file.Close()
+
+	var auths dockerConfigFile
+	fStat, err := os.Stat(path)
+	if err != nil {
+		return dockerConfigFile{}, errors.Wrapf(err, "error getting stat of file %q", path)
+	}
+	if fStat.Size() > 0 {
+		raw, err := ioutil.ReadFile(path)
+		if err != nil {
+			return dockerConfigFile{}, errors.Wrapf(err, "error reading file %q", path)
+		}
+
+		if err = json.Unmarshal(raw, &auths); err != nil {
+			return dockerConfigFile{}, errors.Wrapf(err, "error unmarshaling JSON at %q", path)
+		}
+	} else {
+		newMap := make(map[string]dockerAuthConfig)
+		auths.AuthConfigs = newMap
+	}
+	return auths, nil
+}
+
+// addAuthToJSON either adds a new registry url with its encoded user:pass
+// or modifies an existing registry url
+func addAuthToJSON(url, auth, path string) error {
+	file, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0755)
+	if err != nil {
+		return errors.Wrapf(err, "error creating/opening file %q", path)
+	}
+	defer file.Close()
+
+	auths, err := readJSONFile(path)
+	if err != nil {
+		return errors.Wrapf(err, "error reading JSON file %q", path)
+	}
+
+	enAuth := dockerAuthConfig{Auth: auth}
+	auths.AuthConfigs[url] = enAuth
+
+	newData, err := json.MarshalIndent(auths, "", "\t")
+	if err != nil {
+		return errors.Wrapf(err, "error marshaling JSON %q", path)
+	}
+
+	if err = ioutil.WriteFile(path, newData, 0755); err != nil {
+		return errors.Wrapf(err, "error writing to file %q", path)
+	}
+	return nil
+}
+
+// encodeUserAndPass returns the base64 encoding of user:pass
+func encodeUserAndPass(username, password string) string {
+	return base64.StdEncoding.EncodeToString([]byte(username + ":" + password))
```

Comment:
This is a not-that-complex one-liner with a single user; the code would overall be shorter if it were inlined into the caller.

Reply:
(Or, perhaps, alternatively, the code could go full OO and have `struct dockerAuth {username, password string}` with a custom `json.Marshaller` and `json.Unmarshaller` implementations which completely hide this format from the readers and writers. Very likely not worth the extra complexity.

Or at least place `encodeUserAndPass`+`decodeDockerAuth` together, and have them named consistently, so that the overhead of the one-liner function pays for itself in consistency.)

Reply:
inlined into caller